### PR TITLE
Solve the HiDPI issues of the setup dialog (#2510)

### PR DESCRIPTION
**Pull request by michaelgregorius**
This pull request solves the HiDPI issues of the setup dialog described in #2510 by using Qt widgets and layouts. The Qt widgets solve the HiDPI issue as they scale according to the resolution. The layouts enable to make the dialog resizeable.

The "Autosave" and "Buffer size" tabs have been slightly redesigned so that the revert/reset button is next to the slider. See the images below.

The audio driver settings dialogs are now also shown in a QGroupBox. They now use a `QFormLayout` to layout their labels and input widgets. The same applies to the MIDI setup dialog. Doing so gives a much tidier look than the current implementation with hard-coded sizes and positions

## How does it compare to the existing dialogs?
LMMS-Setup-New-General
LMMS-Setup-Old-General
LMMS-Setup-New-Performance
LMMS-Setup-Old-Performance
LMMS-Setup-New-Audio
LMMS-Setup-Old-Audio
LMMS-Setup-New-MIDI
LMMS-Setup-Old-MIDI
LMMS-Setup-New-Paths
LMMS-Setup-Old-Paths

## Technical details
Setting a fixed size has been removed from the setup dialog so that it can be resized.
    
The settings widget (`settings_w`) now has a layout to which the sub dialogs are added. This is done so that the layout sizes are propagated upwards, i.e. all widgets along the chain now have a layout so that size hints are determined correctly.
    
Instances of `TabWidget` are replaced with instances of `QGroupBox`. Each group box has a layout to which its children, e.g. check boxes, are added. These group boxes are then added to the layouts of their parents. Doing so also removes the need to count how many instances have been added and calculations on where to put the children.
    
Instances of `LedCheckBox` are replaced with instances of `QCheckBox`. This is done in the new helper lambda `addCheckBox`. It's very similar to the previously used `addLedCheckBox` but creates a `QCheckBox` instead of a `LedCheckBox`. It returns the created `QCheckBox` as a result. If a layout is provided the created check box is also added to the layout before it is returned.
    
The helper lambda `addPathEntry` has been adjusted to create a `QGroupBox` and to put all its children in a layout.

`AudioDeviceSetupWidget` now inherits from `QGroupBox` instead of `TabWidget` so that the label is shown in a readable font size.

## Note
The layouts of the setup widgets for `AudioSndio` and `AudioSoundIo` have been adjusted "blindly" because I could not compile the project with them enabled.

### CODE REVIEW

**Comment by tresf** [comment]
> What was the force push about? 

Sorry, it was a bad push to master, had to hard reset to wipe out the changes.

> It seems that you have forcefully removed commit [c53f3d9]

Yeah I guess I did.  that wasn't intended.  It's back now.

> I think the correct way forward would be to set master back to https://github.com/LMMS/lmms/commit/c53f3d981b9710bb1601e9213f925d95bdd9bd72

Done.

> and to undo the changes of https://github.com/LMMS/lmms/commit/c53f3d981b9710bb1601e9213f925d95bdd9bd72 with yet another commit.

I didn't mean to undo it.

**Comment by tresf** [comment]
Hmmm... @michaelgregorius CMT is still listing in this PR (and I assume others).  I'm not sure what to do to fix this.  I was pretty sure that I had reset master properly but this submodule is giving me grief.

**Comment by Rossmaxx** [comment]
>  I was pretty sure that I had reset master properly

Did you run sobmodule update?

**Comment by michaelgregorius** [comment]
> Hmmm... @michaelgregorius CMT is still listing in this PR (and I assume others). I'm not sure what to do to fix this. I was pretty sure that I had reset master properly but this submodule is giving me grief.

Hm, that's strange. The commits here show c53f3d981b9 again. I assume you did something similar to `git reset master c53f3d981b9` (don't know the exact syntax from the top of my head) and simply pushed the changes?

Perhaps something will change once the commit is verified. The checks are still running as of now.

**Comment by tresf** [comment]
> > Hmmm... @michaelgregorius CMT is still listing in this PR (and I assume others). I'm not sure what to do to fix this. I was pretty sure that I had reset master properly but this submodule is giving me grief.
> 
> Hm, that's strange. The commits here show [c53f3d9](https://github.com/LMMS/lmms/commit/c53f3d981b9710bb1601e9213f925d95bdd9bd72) again. I assume you did something similar to `git reset master c53f3d981b9` (don't know the exact syntax from the top of my head) and simply pushed the changes?

Correct.

**Comment by michaelgregorius** [comment]
Commit c53f3d981b9 is now verified and it still shows in the list of commits in this PR (and I assume also in the other ones).

I guess the difference is that the commit was removed with a force push but reintroduced with a normal push. I wonder if GitHub would have reevaluated all PRs if the fix had also been force pushed.

File: data/themes/default/style.css
Comment (zonkmachine) [inline data/themes/default/style.css:7]: We also maintain the **classic** theme. The same code is present there. As I understand it, this regards only the text with the checkboxes in **Settings**.

**Review by he29-net — COMMENTED** [review]
Looks good; it may have lost some of the "LMMS character", but proper scaling and responsiveness should probably come first. Apart from the open comments, the I only small code-related (non-)issue I noticed is that some of the modified or added includes are not in alphabetical order.

As for scaling, I tried running with `QT_AUTO_SCREEN_SCALE_FACTOR=0` (top) to see how it could look on a low DPI screen with no scaling, compared to how LMMS runs on my machine natively (bottom; Debian, Xfce, 155 DPI):
screen

The only thing that seems a bit off is the size of OK and Cancel buttons (they have nearly no margins in the non-scaled version). But some other things look wonky as well (Song-Editor buttons for instance), so maybe this override is not completely accurate. I guess I should find a different machine to get a better reference.

---
A few more notes on issues not necessarily introduced by this PR; if it's not too much work they could be addressed here, or moved to a new issue:

 - `AudioAlsaSetupWidget` (and other selectors in the setup dialog) use a standard `QComboBox`, but portaudio and soundio use automatable LMMS `ComboBox` which probably does not make much sense;

 - tabs General and Paths have an outer border around them (for me it's visible only on the top and left sides), the others do not have it and look much cleaner IMO;

 - unlike `TabWidget`s, `QGroupBox`es have different background color (window background is #262b30, group boxes are #262a2f); the only other place where I found this combination is LADSPA browser, so I'm not sure if it's intentional, or a theme issue that remained hidden until now (because `QGroupBox` was so rarely used);

 - looking at "HQ mode for output audio device", does it even _do_ anything? I don't see its model `m_hqAudioDev` used anywhere else but the setup dialog.

**Comment by irrenhaus3** [comment]
Tested this and loved it, but there are two things that should be looked at:
1) Since the dialog is now resizable, its minimum size should be considered. As of now, it can be squashed down to this size 
image
which is not very useful. Using the previous hard-coded size as the new minimum size would work here.
    
    


2) The reset buttons next to the Autosave and Audio Buffer size interval sliders look like this: 
image
The bitmap icon doesn't scale well, which is out of scope for this PR; but perhaps the button should remain square-shaped as it was before, so the icon can be centered with enough vertical margin around it.

**Comment by he29-net** [comment]
@irrenhaus3 The big left icons all look pretty broken as well, so the size of the reset button probably isn't the issue. Are you (or your window manager) forcing Qt to do fractional scaling by any chance? All the icons look like they were scaled by 1.2 (or some similar non-integer amount) and using very bad interpolation.

In my experience Qt 5 likes to scale UI elements either at 100 or 200 % and nothing between, unless explicitly instructed to (which LMMS widgets do not do). So running on high DPI screen should result in sharp text (rendered natively) and pixelated bitmaps (scaled to 200 % without interpolation to keep them looking sharp) like this:
screen

I have no clue how to achieve a proper high-DPI support, when every platform and every desktop environment (and every other Qt version) seem to be doing things differently. Maybe it would be best to ignore any inputs from the system and add an option for user-defined scaling to the settings...

**Comment by irrenhaus3** [comment]
@he29-net It's a KDE plasma desktop running on X11, which according to its own display configuration manager limits scaling factors to multiples of 6.25%. So it's pretty believable that it enforces fractional scaling. The screenshots were taken at a scaling factor of 125%, so your 1.2 estimate was pretty close. :D
But yeah, in this case there's probably nothing to be done here for now, that issue is definitely out of scope for this PR.

**Comment by Spekular** [comment]
Testing on Win10 on two different screens*, this is more legible than the existing dialogue at every scale setting I've tried. Being able to resize is a nice plus, as is the removal of all the manual `move`s in the code. Unfortunately text and icons are still pixelated at integer scales and blurry at fractional scales, but this is no worse than it was before.

*smaller 2160p monitor normally used at 200% scale, bigger 1440p monitor normally used at 100% scale

File: data/themes/default/style.css
Comment (michaelgregorius) [reply data/themes/default/style.css:7]: @zonkmachine, with the classic theme the change has no effect. It always looks as follows:

2510-SetupDialogClassic

File: src/gui/modals/SetupDialog.cpp
Comment (michaelgregorius) [reply src/gui/modals/SetupDialog.cpp]: Fixed with commit d12675af2f19dd999bfe8f9ab249f21b7171cbd2.

File: src/gui/modals/SetupDialog.cpp
Comment (michaelgregorius) [reply src/gui/modals/SetupDialog.cpp]: Done with commit a54dc9a05f2ff73ea6c4f83012e9ec8a980a6979.

**Comment by michaelgregorius** [comment]
Hi @he29-net, @irrenhaus3,

most of your remarks have been addressed with the commits that have just been added.

## Remarks by @he29-net

> Looks good; it may have lost some of the "LMMS character", but proper scaling and responsiveness should probably come first. 

Part of the character can be brought back once pull request #2068 is merged. It extends the LED check box with the option to use non-scaled text. See the later screenshots in the PR. Once it's in the `QCheckBox` instances could be replaced with the LMMS checkbox again.

> [...] the I only small code-related (non-)issue I noticed is that some of the modified or added includes are not in alphabetical order.

Fixed with commit 85310e75d3d6e658d3f7c101de884dd5194eb910 (only the stuff that I have touched).

> `AudioAlsaSetupWidget` (and other selectors in the setup dialog) use a standard `QComboBox`, but portaudio and soundio use automatable LMMS `ComboBox` which probably does not make much sense;

Yes, this should be moved into a new issue as it is not directly related to the nature of this pull request.

> tabs General and Paths have an outer border around them (for me it's visible only on the top and left sides), the others do not have it and look much cleaner IMO;

I have fixed this with commits cc6a0e86b5942c90b4b5c538cbcd8227faaec60a and 1399d5906f06d2f384e873abb1f5b0716f64ef02.

> unlike `TabWidget`s, `QGroupBox`es have different background color (window background is #262b30, group boxes are #262a2f); the only other place where I found this combination is LADSPA browser, so I'm not sure if it's intentional, or a theme issue that remained hidden until now (because `QGroupBox` was so rarely used);

This seems to be some strange Qt quirk. If you check the title of the group box you will find that it's #262b30. The content has #262a2f. If you change the background of the `QGroupBox` to something completely different, e.g. green, you will find that the title and the content are green but the colors are still different. :(

> looking at "HQ mode for output audio device", does it even _do_ anything? I don't see its model `m_hqAudioDev` used anywhere else but the setup dialog.

This would also be something for another issue.

> The only thing that seems a bit off is the size of OK and Cancel buttons (they have nearly no margins in the non-scaled version). But some other things look wonky as well (Song-Editor buttons for instance), so maybe this override is not completely accurate.

This might be a Qt render thing as I cannot remember to have changed anything with regards to the "OK" and "Cancel" buttons.

## Remarks by @irrenhaus3
> Since the dialog is now resizable, its minimum size should be considered.

Fixed with commit b29a46edf8d27106d7e43b0b3c27d7edce5554b3.

> [...] but perhaps the button should remain square-shaped as it was before, so the icon can be centered with enough vertical margin around it.

Fixed with commit 4f94c3b13cec0b5c2326bc9ce08402f72fb0b2ac.

**Comment by michaelgregorius** [comment]
Hi  @zonkmachine, @he29-net, I'd like to request a final review so that we can also get this HiDPI change in.

**Comment by zonkmachine** [comment]
I've had an issue with some messages showing up underneath gui elements in the Settings window which is fixed by this PR. I assume the reason for the fix is that the window now is allowed to expand downwards. The new window is noticeably larger on my screen/setting which is 1920x1080. I think this looks really good.

hdpipre1
hdpipre2
hdpipre3
hdpipre4
hdpipre5

**Review by zonkmachine — APPROVED** [review]
Based on testing. Approved.

**Review by he29-net — APPROVED** [review]
Changes look good to me, so as far as I'm concerned this should be ready to go. :+1:

**Comment by michaelgregorius** [comment]
Thanks, @zonkmachine and @he29-net!

Here's how it could look once #2068 is merged. With the blue LEDs like in master:

Screenshot_20231001_220821

Alternative with green LEDs:

Screenshot_20231001_220612

**Comment by JohannesLorenz** [comment]
@michaelgregorius Thanks a lot for your submission! Unfortunatelly, it looks like master got corrupted today: https://github.com/LMMS/lmms/commit/339ee6d1e9db5980c15566de911ff8e076313a1e changes more than 400 files, including even translations and changes that actually have been done. This causes merge issues on many current branches. Also, the commit message does not explain much about those 400 files changed.

I hope you are OK if we force-remove all master commits from today? Or was that really intended?

**Comment by JohannesLorenz** [comment]
By the way, I am not sure if the issue come from this PR or another one, but I did not know how to contact you.

**Comment by JohannesLorenz** [comment]
Sorry. I just got confused why that master commit had changes from 400 files in the diff, which were unrelated to your PR. But it seems like you first merged master into your branch and then did an FF merge of your branch into master, resulting in github showing a master commit with 400 changes that already are on master.

So it is a bit misleading, but still OK.

**Comment by michaelgregorius** [comment]
Hi @JohannesLorenz,

where does it show the large diff? If I click on the merge commit 8b94bd8e654 it only shows 17 changed files.

Is this perhaps something tool related? If I check the merge commit with "gittyup" it also only shows 17 changed files:

Screenshot_20231002_112518

All I can say is that I have only used "legal" git operations like branching off, committing changes and merging master to resolve conflicts. For the latter I have used a temporary "merge branch" that was later forwarded into the official branch. However from my understanding this should have the same effect as if I would have done everything in the original branch in the first place (due to the fast-forward).

**Comment by he29-net** [comment]
> If I check the merge commit with "gittyup" it also only shows 17 changed files

I think the issue is that https://github.com/LMMS/lmms/commit/339ee6d1e9db5980c15566de911ff8e076313a1e shows up at https://github.com/LMMS/lmms/commits/master, along with all the other commits from your branch. So the main LMMS commit history has the `339ee6..` commit, which claims to be making a lot of changes to `master` that make no sense, because the changes were already present in `master`.

I asked about it in the Discord dev channel yesterday, but I still don't think I fully understand how it happens. I just know that "GitHub PR merge" can use has three different methods, and none of them is a simple "git merge" as you would type it in CLI. I.e., in this case, instead of adding one commit to LMMS master, it transplanted the whole commit history, plus a merge commit that just says "a merge happened" and repeats the final changes that result from the previously added commit history. No idea how that works, since all the changes look to me like as if they were done twice.

I guess I'll need to carefully read https://docs.gitlab.com/ee/user/project/merge_requests/methods/ and look into all the related terms (assuming GitLab does it the same way). Hopefully that will make it a bit clearer to me. It's fascinating to me how, after using git for years, I still keep running into a completely new territory from time to time... :)

**Comment by michaelgregorius** [comment]
Yes, there is always something new to learn with Git. I enjoy the theory that Git has become so widely used and successful among developers exactly because it is not easy to use but because it's another complex tool for them to master and where you can develop lots of expertise. 😄

Another reason why the history might look "wonky" with regards to this pull request might be the force push that was done by @tresf about three months ago. It was discussed in this issue starting from [here](https://github.com/LMMS/lmms/pull/6779#issuecomment-1652100862). Above the comment you can find the event as reported by Github.

If I remember correctly my branch was originally branched off of commit c2ae07a0877 which was forcefully removed with a force push where master was reset to its parent commit 5ccc8d952ac. At the time when this was done the removed commit became part of my pull request because to Git it now seems to have looked as if I had started from 5ccc8d952ac and wanted to add c2ae07a0877 as part of my changes. For some reason it does not look like this anymore and I don't know how this was resolved. @tresf, can you help here?

Force pushes should almost never be used especially in a shared project like this one. Problems should always be fixed with forward commits and not by forcefully removing commits with force pushes. The reason is that doing so causes problems for developer who have already pulled these commits and base their work on them.

Rewriting of Git history should only be done in local repositories that do not have any changes pushed to public origins yet. And here the rewriting must not cross the border into shared Git history as well if the changes are later intended to be pushed into the public.

**Comment by tresf** [comment]
> For some reason it does not look like this anymore and I don't know how this was resolved. [@tresf](https://github.com/tresf), can you help here?

No clue.

> Force pushes should almost never be used especially in a shared project like this one.

They are almost never used.  Project maintainers need to temporarily disable branch protection to do this, and this is what was done, and it was communicated.  Force-pushes are sane in moderation.  I know project maintainers often don't agree with this, but usually it's because force pushes are used to rewind history that the world has already consumed.  This happened to us with the [Exprtx submodule](https://github.com/LMMS/lmms/issues/5064).  The author wiped out our submodule reference, breaking all tags of LMMS that relied upon it.  In regards to the commit that caused problems, this wasn't the case and the hard reset was a -- relatively sane rollback.  Due to the havoc it created... in hindsight... I should have reverted the submodule in a subsequent commit.

That said, this PR nearly triggered the same action.  @PhysSong was ready to roll back `master` branch. 💀 

image

**Comment by tresf** [comment]
> For some reason it does not look like this anymore and I don't know how this was resolved. @tresf, can you help here?

> > No clue.

I actually mean this, BTW.  I don't know why the dangling commits appeared in all open PRs and I don't know why they disappeared.  It's a mystery to me. 🍻

**Comment by michaelgregorius** [comment]
Thanks for chiming in @tresf! I can only explain why they appeared but not why they disappeared. 😅

I think as long as no other pull requests are based on the removed commit this problem should not reappear in other pull requests.

**Comment by tresf** [comment]
> "They are almost never used." and the first reaction being "Let me do a force push".

Quantifying this: 8,000 commits over about 12 years: It's happened about 10 times.

This PR was particularly alarming because the commits get woven into master's history, which we normally avoid.

**Comment by michaelgregorius** [comment]
By the way, it looks like c53f3d981b9 is still/back in the repository, @tresf. So you might want to check if the cmt submodule looks as expected.

**Comment by michaelgregorius** [comment]
Ok, I have now taken some time to dig through gitk and to check the Git history.

## What you are seeing in commit 339ee6d1e9d is perfectly ok
The changes in commit 339ee6d1e9db5980c15566de911ff8e076313a1e look perfectly ok and as expected. It's important to understand that what you are seeing in this commit is not a diff to master but the diff from master to my development branch at that time.

So the alarm was a false alarm that was caused by a matter of misinterpretation of what's being shown in the diff. Nothing is corrupted.

## A more detailed look
I have branched off at c53f3d981b9710bb1601e9213f925d95bdd9bd72 and master then continued on it's own merry path with a9d49d4ff70642c958380a0b2e4fc6180c1a1877 being the next commit. This should have happened around 16.07. because that's when my first commit in the "Setup Dialog" branch has happened.

Over the next few months the master branch then continued to receive merges and new commits from other pull requests, i.e. lots of development took place in it in the meantime.

The first merge of master into my branch was done on 01.10. via 339ee6d1e9db5980c15566de911ff8e076313a1e.  This merged two and a half months of development from master into my dev branch so it expected that lots of files show up in the diff because my branch has never seen these changes and they are all added to my branch because I am syncing it with master. So what you are seeing is all the work that has happened between around 16.07. and 01.10. on master.

Put differently: the longer a development branch is running without being merged into master the more changes are accumulating in master. If you then merge all these changes very late from master into the development branch you will see all the work that has taken place in the meantime.

So everything is fine. And thanks for not nuking my work with a force push. 😉

## Proposal for a procedure if a force push is done
In case you aren't already doing so: if you do a force push the next time please create a branch of the current state to save it. I have not deleted the branch in my remote or locally so I could have reinstated the changes if they would have been removed by a force push based on false assumptions. However, if someone else is quicker with deleting everything on their side because they deem their changes safe now that they are in master their work might be gone though (or distributed in the repositories of other contributors if they are lucky).

In some cases the work might be reinstated through the ref log but I think doing so is not fun either.

**Comment by JohannesLorenz** [comment]
So, the state is not corrupted. There have been cases of corruption though, where we did an instant force-push. Sometimes you have to decide whether you want to accept a master commit which is not at all doing what it should and/or increases the complexity of the history and the time every further user will waste when they clone master. Luckily, this was not the case with your commit. But in such case, it can be good to do a force push ASAP, in order to avoid other users getting trouble with their history.

Anyways, while it is "perfectly OK" from a "corruption" point of view, some would not consider it "perfectly OK" from a "cleanness" point of view. Git has revolutionized the way we merge code. Traditional merge commits are often unpopular for a couple of reasons, let me show this with a picture of the current situation:

image

From the 14 commits you committed on master (they are all on master now and all part of master's history, because in Git, master is not a set of commits, but only a pointer to one commit), you added

* 4 merge commits
* At least 2 fixup commits (or commits containing fixups)

The `git log` now gets longer that it would need to be by at least 5 commits. All this can be avoided by

* Squashing all commits into one
* Doing "git rebase -i" before you commit to order and squash commits on a granular level, then do a "rebase + ff" merge

Some people may still argue that merge commits on master are fine (or at least in some cases), though I think the consensus is to avoid them. However, fixup commits on master are generally seen as bad style.

I have done the same things in the past. There are still fixup commits and (very) strange branch patterns from myself on zyn's master from 2017. I genuinely value your skills as a developer, especially your commitment to providing detailed commit messages and the fact that these are all free time contributions. So don't get me wrong, this is rather a summary how most developpers prefer to merge, not an attack :slightly_smiling_face: .

**Comment by michaelgregorius** [comment]
> So don't get me wrong, this is rather a summary how most developpers prefer to merge, not an attack 🙂 .

No offense taken. I guess in part it's just what one is used to.

At work we always do merge commits and we are definitively not a small team. I can also assure you that the few additional commits in the history do not have much of an impact on checkout performance. We have a much larger code base than LMMS at work and the checkouts are still quick.

One advantage of doing merge commit is that they preserve the full history (as intended by the authors) of how a feature came to be and that they for example allow you to bisect on a much more detailed scale in case it becomes necessary to find out what change exactly introduced a bug. I think this is also one of the reasons why there are also people who advocate to do many small commits.

> From the 14 commits you committed on master (they are all on master now and all part of master's history, because in Git, master is not a set of commits, but only a pointer to one commit), you added
> 
>     * 4 merge commits
> 
>     * At least 2 fixup commits (or commits containing fixups)
> 
> 
> The `git log` now gets longer that it would need to be by at least 5 commits.

This is more or less just a problem of presentation. Github presents the log by sorting all commits chronologically regardless of where they came from. This can look messy and noisy.

However, if you look at the log as a graph you can always follow the first parents of the master branch and will only see the events that took place on master, i.e. fast forward commits or final merge commits.

The following command for example shows the "clean" history of master without showing any commits from development branches:
```
git log --first-parent --pretty=format:"%Cblue %h %Cgreen %an %Creset %s" --graph
```

We can also show only the merge commits that have gone into master:
```
git log --first-parent --pretty=format:"%Cblue %h %Cgreen %an %Creset %s" --graph --merges
```

> All this can be avoided by
> 
>     * Squashing all commits into one
> 
>     * Doing "git rebase -i" before you commit to order and squash commits on a granular level, then do a "rebase + ff" merge

I think this approach will not work if there are merges from master in between. Or is it possible to squash merge commits? I cannot imagine how this should work.

If git does not support squashing merge commits then the proposed flow to clean everything up would likely become pure hell for long running development branches like for example the one from #2068.

The reason is that you cannot simply rebase very old commits onto the newest code because the code has usually diverged so much that every rebase will give you tons of merge conflicts which might not even be resolvable because the old code sees a completely different environment now.

Rebasing can also lead to problems of it's own. In the end a rebase can be interpreted as: "Let's pretend my work did not start on that commit but on another one." This can create the following problem:
* Let's imagine some dev branch D which branched off of commit A.
* In that branch we commit some code that calls a certain function with certain parameters. The code compiles and works and we create a commit W.
* Now we rebase commit W on a later commit B. In that commit the signature of the function has changed.
* The rebase will go through (because git does not compile) but the code from the new commit will not compile anymore.

The aforementioned problem will for example cause problems when bisecting the code. This becomes even more complicated if several commits are rebased. The only way to solve this would be to fix everything with a fix-up commit and squash everything right up to the first rebased commit. Again throwing away valuable information.

The alternative flow is to merge B into W, fix the build and then finalize the merge. Now there are only working commits and all information is still retained.

> Some people may still argue that merge commits on master are fine (or at least in some cases), though I think the consensus is to avoid them. However, fixup commits on master are generally seen as bad style.

What follows is just my opinion: if the log just looks like a flow of single fast-forward commits then this seems like an attempt to emulate Subversion's history. However, as noted before lots of potentially valuable information is thrown away by doing so. And it also introduces lots of efforts while throwing away valuable information.

To me the power of Git merges lies in the fact that you can see a detailed history of what got merged into which on a detailed level. And that it retains a history of the merges in the first place. If I remember correctly, in Subversion this information is more or less lost.

> I genuinely value your skills as a developer, especially your commitment to providing detailed commit messages and the fact that these are all free time contributions.

Thanks! 😊

**Comment by JohannesLorenz** [comment]
> The following command for example shows the "clean" history of master without showing any commits from development branches:
> 
> ```
> git log --first-parent --pretty=format:"%Cblue %h %Cgreen %an %Creset %s" --graph
> ```

OK, this is really cool for this situation. I can even run a `git log --first-parent`, and it seems to then show the exact master's history. However, it will not work for all cases. If you first merge master into your branch and only then merge your branch ff-only into master, the option will take the path of your side branch.

> We can also show only the merge commits that have gone into master:
> 
> ```
> git log --first-parent --pretty=format:"%Cblue %h %Cgreen %an %Creset %s" --graph --merges
> ```
> 
> > All this can be avoided by
> > ```
> > * Squashing all commits into one
> > 
> > * Doing "git rebase -i" before you commit to order and squash commits on a granular level, then do a "rebase + ff" merge
> > ```
> 
> I think this approach will not work if there are merges from master in between. Or is it possible to squash merge commits? I cannot imagine how this should work.

In github, there is the "Squash" option when merging. This always seems to work, even if you have merge commits on your branch. It is probably just running `git merge --squash`. I assume this command does nothing else than checking out your branch, then runs `git reset master` and then does a commit.
 
> If git does not support squashing merge commits then the proposed flow to clean everything up would likely become pure hell for long running development branches like for example the one from #2068.
>
> The reason is that you cannot simply rebase very old commits onto the newest code because the code has usually diverged so much that every rebase will give you tons of merge conflicts which might not even be resolvable because the old code sees a completely different environment now.

Yes, I remember that problem, and in such cases, merges can be helpful. Or partial squashing `git rebase -i` of your sidebranch to reduce your hell prior to rebasing (in case it applies).

For `lv2-worker`, I ended up with ~35 commits, of which were 5 merge commits. I wanted to partially squash, but not full squash, because the branch had around 5 different topics that it touched, which could be represented atomically. `git rebase -i` did not work due to the merge commits. In this case, I did the reset technique mentioned above, to first squash into one commit, and then `git add -p` to create 5 commits out of it. This was a bit difficult, but I did not know a better technique for this case.

> Rebasing can also lead to problems of it's own. In the end a rebase can be interpreted as: "Let's pretend my work did not start on that commit but on another one." This can create the following problem:
> 
>     * Let's imagine some dev branch D which branched off of commit A.
> 
>     * In that branch we commit some code that calls a certain function with certain parameters. The code compiles and works and we create a commit W.
> 
>     * Now we rebase commit W on a later commit B. In that commit the signature of the function has changed.
> 
>     * The rebase will go through (because git does not compile) but the code from the new commit will not compile anymore.
> 
> 
> The aforementioned problem will for example cause problems when bisecting the code. This becomes even more complicated if several commits are rebased. The only way to solve this would be to fix everything with a fix-up commit and squash everything right up to the first rebased commit. Again throwing away valuable information.

After that workflow, it would look like the dev never looked at commit A, but just did one commit with the B signature in mind. The only information I see which is lost is how the developer first worked with the A signature, but this is OK since A is outdated code.

Well, some developers think that the way (all the commits, fixups etc) they produced their feature branches is still useful after integration to master. Though I think more developers think it is "too much info". Maybe you are referring to that.

**Comment by tresf** [comment]
> Well, some developers think that the way (all the commits, fixups etc) they produced their feature branches is still useful after integration to master. Though I think more developers think it is "too much info". Maybe you are referring to that.

It's too much.

>  first squash into one commit, and then git add -p to create 5 commits out of it. This was a bit difficult, but I did not know a better technique for this case.

This is what I would have expected here if the squash removed too much history.  I agree, it's a bit difficult.

**Comment by michaelgregorius** [comment]
> OK, this is really cool for this situation. I can even run a `git log --first-parent`, and it seems to then show the exact master's history. However, it will not work for all cases. If you first merge master into your branch and only then merge your branch ff-only into master, the option will take the path of your side branch.

Yes, for the first-parent to work the merges always must be from the development branch into master. In you example the development branch would become the first parent in the merge commit and the fast-forward would make it the new master. Hence, the roles would appear reversed.

Unfortunately Git cannot do this any better because as you have already noted the branch names are just pointers to commits and Git has no deeper concepts for them. So Git does not know that "this commit once 'belonged' to the master branch" and "this commit once 'belonged' to the dev branch myFeatureBranch" and that therefore it was the branch "master" that was merged into "myFeatureBranch". This is information is mostly lost on merge commits.

Hence, it's a question of workflow and Github available options make sure that the merges are consistent.

> [...] It is probably just running `git merge --squash`. I assume this command does nothing else than checking out your branch, then runs `git reset master` and then does a commit.

Today I learned. Thanks! 🙂

> After that workflow, it would look like the dev never looked at commit A, but just did one commit with the B signature in mind. The only information I see which is lost is how the developer first worked with the A signature, but this is OK since A is outdated code.

The thing is just that if you rebase several commits from a branch they will just be applied starting from the new base commit. The rebase will only check for merge conflicts but in the example given there would be none. However, the code would not compile. So in the case where people want to keep all commits this will create a problem because a commit that once compiled and worked would be broken now.

> Well, some developers think that the way (all the commits, fixups etc) they produced their feature branches is still useful after integration to master. Though I think more developers think it is "too much info". Maybe you are referring to that.

Yes, it can indeed be useful, especially if you commit many small changes. If you have to find out what caused a bug in master that was introduced with a merge commit then the bisection will also look through all the individual small commits. Because they all compile as they have never been rebased you can find the exact change that introduced the problem.

If you merge with squash and rebase you will only know that the one big fat commit with potentially tons of changes has introduced the bug.

**Comment by JohannesLorenz** [comment]
> Today I learned. Thanks! 🙂

Np. I think I learned most of what I know now here at LMMS (and ZynAddSubFX) :smiley:

**Pull request merged**
